library/src-bootstrapped/scala/Tuple.scala

```diff
@@ -0,0 +1,278 @@
+   *  The result is typed as `(F[A1], ..., F[An])` if the tuple type is fully known.
+   *  If the tuple is of the form `a1 *: ... *: Tuple` (that is, the tail is not known
+   *  to be the cons type.
+   */
+  inline def map[F[_]](f: [t] => t => F[t]): Map[this.type, F] =
+    runtime.Tuples.map(this, f).asInstanceOf[Map[this.type, F]]
+
+  /** Given a tuple `(a1, ..., am)`, returns the tuple `(a1, ..., an)` consisting
+   *  of its first n elements.
+   */
+  inline def take[This >: this.type <: Tuple](n: Int): Take[This, n.type] =
+    runtime.Tuples.take(this, n).asInstanceOf[Take[This, n.type]]
+
+
+  /** Given a tuple `(a1, ..., am)`, returns the tuple `(an+1, ..., am)` consisting
+   *  all its elements except the first n ones.
+   */
+  inline def drop[This >: this.type <: Tuple](n: Int): Drop[This, n.type] =
+    runtime.Tuples.drop(this, n).asInstanceOf[Drop[This, n.type]]
+
+  /** Given a tuple `(a1, ..., am)`, returns a pair of the tuple `(a1, ..., an)`
+   *  consisting of the first n elements, and the tuple `(an+1, ..., am)` consisting
+   *  of the remaining elements.
+   */
+  inline def splitAt[This >: this.type <: Tuple](n: Int): Split[This, n.type] =
+    runtime.Tuples.splitAt(this, n).asInstanceOf[Split[This, n.type]]
+}
+
+object Tuple {
+
+  /** Type of the head of a tuple */
+  type Head[X <: NonEmptyTuple] = X match {
+    case x *: _ => x
+  }
+
+  /** Type of the tail of a tuple */
+  type Tail[X <: NonEmptyTuple] <: Tuple = X match {
+    case _ *: xs => xs
+  }
+
+  /** Type of the concatenation of two tuples */
+  type Concat[X <: Tuple, +Y <: Tuple] <: Tuple = X match {
+    case EmptyTuple => Y
+    case x1 *: xs1 => x1 *: Concat[xs1, Y]
+  }
+
+  /** Type of the element a position N in the tuple X */
+  type Elem[X <: Tuple, N <: Int] = X match {
+    case x *: xs =>
+      N match {
+        case 0 => x
+        case S[n1] => Elem[xs, n1]
+      }
+  }
+
+  /** Literal constant Int size of a tuple */
+  type Size[X <: Tuple] <: Int = X match {
+    case EmptyTuple => 0
+    case x *: xs => S[Size[xs]]
+  }
+
+  /** Fold a tuple `(T1, ..., Tn)` into `F[T1, F[... F[Tn, Z]...]]]` */
+  type Fold[T <: Tuple, Z, F[_, _]] = T match
+    case EmptyTuple => Z
+    case h *: t => F[h, Fold[t, Z, F]]
+
+  /** Converts a tuple `(T1, ..., Tn)` to `(F[T1], ..., F[Tn])` */
+  type Map[Tup <: Tuple, F[_]] <: Tuple = Tup match {
+    case EmptyTuple => EmptyTuple
+    case h *: t => F[h] *: Map[t, F]
+  }
+
+  /** Converts a tuple `(T1, ..., Tn)` to a flattened `(..F[T1], ..., ..F[Tn])` */
+  type FlatMap[Tup <: Tuple, F[_] <: Tuple] <: Tuple = Tup match {
+    case EmptyTuple => EmptyTuple
+    case h *: t => Concat[F[h], FlatMap[t, F]]
+  }
+
+  /** Filters out those members of the tuple for which the predicate `P` returns `false`.
+   *  A predicate `P[X]` is a type that can be either `true` or `false`. For example:
+   *  ```
+   *  type IsString[x] = x match {
```

Comment:
As pointed out here https://github.com/lampepfl/dotty/issues/10896 this example is misleading and won't work without a type bound
```scala 
type IsString[x] <: Boolean = x match {
```

Reply:
Fixing that is independent of this PR.